be/src/exec/olap_scan_prepare.cpp

```diff
@@ -424,7 +424,7 @@ requires(!lt_is_date<SlotType>) Status ChunkPredicateBuilder<E, Type>::normalize
             ValueType value;
             ASSIGN_OR_RETURN(auto* expr_context, _exprs[i].expr_context(_opts.obj_pool, _opts.runtime_state));
             bool ok =
-                    get_predicate_value<Negative>(_opts.obj_pool, slot, root_expr, expr_context, &value, &op, &status);
+                    get_predicate_value<Negative>(_opts.obj_pool, slot, expr_context->root(), expr_context, &value, &op, &status);
```

Comment:
Should use `get_root_expr(expr_context->root())` instead of `expr_context->root()`.

Maybe it's better to modify `BoxedExpr::root()` as follows and call `_exprs[i].root()` instead of `expr_context->root()`.
```cpp
Expr* BoxedExpr::root() const {
    if (new_expr_ctx == nullptr) {
        return get_root_expr(root_expr);
    } else {
        return get_root_expr(new_expr_ctx);
    }
}
```

Reply:
> Should use get_root_expr(expr_context->root()) instead of expr_context->root().

ok, will update


> Maybe it's better to modify BoxedExpr::root() as follows and call _exprs[i].root() instead of expr_context->root().

i think it will be more error prone, because need to keep in mind about side effects of `expr_context`
and `root()` will return different results before and after `expr_context()`

Reply:
> i think it will be more error prone, because need to keep in mind about side effects of expr_context
and root() will return different results before and after expr_context()

Make sense. How about making `expr_context()` return `std::pair<ExprContext*, Expr*> [ctx, root]` instead of `ExprContext*`? Then we can use this new `root` and `ctx`.